plugins/wdc/wdc-nvme.c

```diff
@@ -1038,6 +1039,84 @@ static int wdc_create_log_file(char *file, __u8 *drive_log_data,
 	return 0;
 }
 
+static int wdc_get_log(int fd, __u32 nsid, __u8 log_id, bool rae, __u8 uuid_ix,
+		 __u32 data_len, void *data)
+{
+	__u32 offset = 0, xfer_len = data_len;
+	int ret;
+	void *ptr = data;
+
+	/*
+	 * 4k is the smallest possible transfer unit, so by
+	 * restricting ourselves for 4k transfers we avoid having
+	 * to check the MDTS value of the controller.
+	 */
+	do {
+		xfer_len = data_len - offset;
+		if (xfer_len > 4096)
+			xfer_len = 4096;
+
+		ret = nvme_get_log14(fd, nsid, log_id, NVME_NO_LOG_LSP,
+				     offset, 0, rae, uuid_ix, xfer_len, ptr);
+
+		if (ret)
+			return ret;
+
+		offset += xfer_len;
+		ptr += xfer_len;
+	} while (offset < data_len);
+
+	return 0;
+}
+
+static const uint8_t zero_uuid[16] = { 0 };
+
+static int get_wdc_uuid_index(int fd, __u8* uuid_ix, uuid_t uuid)
```

Comment:
This function actually has a bit too much indentation and exceeds the standard 80-column width. I think we can shorten it up a bit if we reorganize the parts. What do you think of the following?

```
static int get_wdc_uuid_index(int fd, __u8* uuid_ix, uuid_t uuid)
{
        struct nvme_id_uuid_list uuid_list = { 0 };
        struct nvme_id_ctrl ctrl;
        int i, ret = -1;

        *uuid_ix = 0;

        /* Determine if the controller supports a UUID List by checking CTRATT */
        ret = nvme_identify_ctrl(fd, &ctrl);
        if (ret)
                return ret;

        if (!(ctrl.ctrattr & NVME_CTRL_CTRATT_UUID_LIST))
                return 0;

        /* UUID list is supported.  Get the WDC UUID entry */
        ret = nvme_identify_uuid(fd, &uuid_list);
        if (ret)
                return ret;

        /* The 0th entry is reserved so start at 1 */
        for (i = 1; i< NVME_MAX_UUID_ENTRIES; i++) {
                struct nvme_id_uuid_list_entry *entry = &uuid_list.entry[i];

                /* The list is terminated by a zero UUID value */
                if (memcmp(entry->uuid, zero_uuid, sizeof(zero_uuid))) {
                        if (memcmp(entry->uuid, &uuid, sizeof(uuid_t)) == 0) {
                                *uuid_ix = i;
                                return 0;
                        }
                } else
                        break;
        }

        return 0;
}
```

Reply:
I see you set the default value for uuid_ix to zero.  Jeff said it should be set to one unless we find a specific uuid entry otherwise I am ok with it..

Reply:
Sure, but the calling function already overrides it to 1 already.

I'm not sure of the rationale for that, though. If you can't find the desired uuid, then the function should abandon the action with an appropriate error.

Reply:
Actually I do see a problem with your suggested code. 
The following line checks to see if we reached the end without finding an entry.  The line after needs to be in the else as follows.
if (memcmp(entry->uuid, zero_uuid, sizeof(zero_uuid))) {
  uuid_ix = 0;
} else {
if (memcmp(entry->uuid, &uuid, sizeof(uuid_t)) == 0) {
/* entry found*/
  uuid_ix = i;
  return 0;
}

Reply:
> Actually I do see a problem with your suggested code.
> The following line checks to see if we reached the end without finding an entry. The line after needs to be in the else as follows.
> if (memcmp(entry->uuid, zero_uuid, sizeof(zero_uuid))) {
> /* entry not found _/ uuid_ix = 0; } else { if (memcmp(entry->uuid, &uuid, sizeof(uuid_t)) == 0) { /_ entry found*/
> uuid_ix = i;
> return 0;
> }

I don't understand. If the entry's uuid doesn't match the zero uuid, why does that mean "entry not found"? Isn't the opposite supposed to be the case?